source/tutorials/nixos-configuration-testing.md

```diff
@@ -0,0 +1,485 @@
+
+The complete `minimaltest.nix` file content looks like the following:
+```{code-block}
+let
+  nixpkgs = <nixpkgs>;
+  pkgs = import nixpkgs {};
+in
+  pkgs.nixosTest {
+    nodes.machine = { config, pkgs, ... }: {
+      boot.loader.systemd-boot.enable = true;
+      boot.loader.efi.canTouchEfiVariables = true;
+
+      services.xserver.enable = true;
+      services.xserver.displayManager.gdm.enable = true;
+      services.xserver.desktopManager.gnome.enable = true;
+
+      users.users.alice = {
+        isNormalUser = true;
+        extraGroups = [ "wheel" ];
+        packages = with pkgs; [
+          firefox
+          thunderbird
+        ];
+      };
+
+      system.stateVersion = "22.05";
+    };
+    testScript = {nodes, ...}: ''
+      machine.wait_for_unit("default.target")
+      machine.succeed("su -- alice -c 'which firefox'")
+      machine.fail("su -- root -c 'which firefox'")
+    '';
+  }
+```
+
+### Running tests
+
+To set up all machines and execute the test script:[^nixdev]
+
+```shell-session
+$ nix-build minimaltest.nix
+```
+
+You'll notice an error message if something goes wrong.[^nixdev]
+
+In case the tests succeed, you should see at the end:[^nixdev]
+
+```shell-session
+...
+test script finished in 10.96s
+cleaning up
+killing machine (pid 10)
+(0.00 seconds)
+/nix/store/bx7z3imvxxpwkkza10vb23czhw7873w2-vm-test-run-unnamed
+```
+
+### interactive python shell to interact with virtual machine
+
+When developing tests or when something breaks, it’s useful to interactively fiddle with the script or access a terminal for a machine.[^nixdev]
+
+To interactively start a Python session with a testing framework:[^nixdev]
+```shell-session
+$ $(nix-build -A driverInteractive minimaltest.nix)/bin/nixos-test-driver
+```
+
+You can run any of the testing operations. The testScript attribute from our minimaltest.nix definition can be executed with `test_script()` function.[^nixdev]
+
+Within this Python shell you can enter a interactive shell you can run python commands like those in the test script.
+
+On the first evaluation of the command the virtual machine is started.
+You can manually trigger the start of the virtual machine by using
+```shell-session
+>>> machine.start()
+```
+for a specific node.
+
+or
+```shell-session
+>>> start_all()
+```
+for all specified nodes.
+
+You can enter a interactive shell on the virtual machine using:
+```shell-session
+>>> machine.shell_interact()
+```
+
+and run commandline commands like:
+```shell-session
+$ uname -a
+Linux server 5.10.37 #1-NixOS SMP Fri May 14 07:50:46 UTC 2021 x86_64 GNU/Linux
+```
+
+
+The nixos manual contains a [reference for commands](https://nixos.org/manual/nixos/stable/index.html#ssec-machine-objects) to design tests.
+
+
+### rerun successful tests
+
+Because test results are stored in the nix-store a successful test is cached, this means that nix will not run the test a second time as long as the test setup (configuration or test script) is not changed.
```

Comment:
```suggestion
Because test results are kept in the Nix store, a successful test is cached.
This means that Nix will not run the test a second time as long as the test setup (machine configuration or test script) stays the same.
```
It may be worthwhile to note that this is true as long as there is no *semantic* change, as a refactoring of the system definition may still produce the exact same configuration.

@Radvendii you might be interested to link your upcoming post here.